dotnet tutorial kiota dependency injection

svrooij opened the pull request:
Fixed #37

svrooij commented:
@baywet here is my tutorial for Kiota with DI solving issue #37

baywet commented on OpenAPI/kiota/toc.yml:
```diff
@@ -77,5 +77,7 @@ items:
   - name: Add an API to search
     href: add-api.md
+  - name: Use Kiota with Dependency Injection in .NET
+    href: tutorials/dotnet-dependencyinjection.md
```
you might want to add an additional dash between dependency and injection here.

svrooij replied:
Will do

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+---
+title: Use Kiota with dependency injection in .NET
+description: Learn how you can combine Kiota with dependency injection in .NET to build API clients.
+author: svrooij
```
add your username to this file please https://github.com/MicrosoftDocs/openapi-docs/blob/main/cspell.json

svrooij replied:
Check 😜
Maybe this should be documented somewhere

baywet replied:
with #88 we're moving away from that for CI anyway

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
````diff
@@ -0,0 +1,288 @@
+Run the following command in the directory you want to create a new project.
+
+```dotnetcli
+dotnet new webapi -o kiota-with-di
````
add webapi to this file please https://github.com/MicrosoftDocs/openapi-docs/blob/main/cspell.json

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+
+## Add dependencies
+
+Before you can compile and run the generated API client, you will need to make sure the generated source files are part of a project with the required dependencies. Your project must have a reference to the [abstraction package](https://github.com/microsoft/kiota-abstractions-dotnet) and some default implementations.
```
```suggestion
Before you can compile and run the generated API client, ensure the generated source files are part of a project with the required dependencies. Your project must reference the [abstraction package](https://github.com/microsoft/kiota-abstractions-dotnet) and default implementations.
```

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
````diff
@@ -0,0 +1,288 @@
+You can then use the [Kiota](/openapi/kiota/install?tabs=bash&wt.mc_id=SEC-MVP-5004985#install-as-net-tool) command line tool to generate the API client classes. We are using the [deserializer](/openapi/kiota/using#--deserializer---ds) and [serializer](/openapi/kiota/using#--serializer---s) options to specify the serialization format, to only use JSON serialization.
+
+```bash
+kiota generate -l csharp -d github-releases.yml -c GitHubClient -n GitHub.ApiClient -o ./GitHub --deserializer Microsoft.Kiota.Serialization.Json.JsonParseNodeFactory --serializer Microsoft.Kiota.Serialization.Json.JsonSerializationWriterFactory
````
I'd leave the defaults for simplicity and remove the serializer and deserializer arguments. Same for the dependencies earlier.
Other documentation pages explore/will explore the serialization mechanisms and the dependencies, and the choices they can make.

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+    /// <summary>
+    /// Adds the Kiota handlers to the service collection.
+    /// </summary>
+    /// <param name="services"><see cref="IServiceCollection"/> to add the services to</param>
```
add cref to this file https://github.com/MicrosoftDocs/openapi-docs/blob/main/cspell.json

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+    /// <remarks>The handlers are added to the http client by the <see cref="AttachKiotaHandlers(IHttpClientBuilder)"/> call, which requires them to be pre-registered in DI</remarks>
+    public static IServiceCollection AddKiotaHandlers(this IServiceCollection services)
+    {
+        services.AddTransient<UriReplacementHandler<UriReplacementHandlerOption>>();
```
out of curiosity here, why not simply call that and feed the array in the services?
https://github.com/microsoft/kiota-http-dotnet/blob/3e2d658c81fa57be2ccfa6d61a7ad2364640ad75/src/KiotaClientFactory.cs#L50

svrooij replied:
This call is initializing them, and I just need the the types to register them in DI.
Or would you say that doing a 

```csharp

var handlers = KiotaClientFactory.CreateDefaultHandlers()
foreach(var handler in handlers) {
  services.AddTransient(typeof(handler));
}
```

is better? Just trying to see what looks better in coding structures that are commonly used.

baywet replied:
The challenge with explicitly enumerating the type is that if tomorrow we add a the middleware handler (let's say something that polls automatically long running operations for example) to the stack, people wouldn't get it automatically.
That's why if we can, it'd be better if it could be more dynamic on that part, even if it means a little more complex code.
does that make sense?

svrooij replied:
I've made them dynamically, but left the old code. I really don't like this new way because it looks really ugly. And it's creating those handlers during startup twice, just to get the type of them.

baywet replied:
assuming this gets implemented, we should be able to clean it up here.
https://github.com/microsoft/kiota-http-dotnet/issues/254

svrooij replied:
I just noticed that the new way does not work because of this one.
https://github.com/microsoft/kiota-http-dotnet/blob/3e2d658c81fa57be2ccfa6d61a7ad2364640ad75/src/KiotaClientFactory.cs#L55

baywet replied:
I believe this was resolved with the update of http. Resolving

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+    /// </remarks>
+    public static IHttpClientBuilder AttachKiotaHandlers(this IHttpClientBuilder builder)
+    {
+        builder.AddHttpMessageHandler<UriReplacementHandler<UriReplacementHandlerOption>>();
```
same here about using the factory?

svrooij replied:
I literally have no clue how you could use an instance of something in a method that expects just the type in the generic specifier.

svrooij replied:
See above

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+using GitHub;
+...
+
+// Add Kiota handlers to the DI container
```
```suggestion
// Add Kiota handlers to the dependency injection container
```

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+}).AttachKiotaHandlers();
+
+// Register the GitHub client
+builder.Services.AddTransient(sp=> sp.GetRequiredService<GitHubClientFactory>().GetClient());
```
```suggestion
builder.Services.AddTransient(sp => sp.GetRequiredService<GitHubClientFactory>().GetClient());
```

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+
+The `GitHubClientFactory` class is responsible for creating the client, and it is also responsible for creating the `IAuthenticationProvider`. In this sample, we are using the `AnonymousAuthenticationProvider`, which is a simple provider that does not add any authentication headers. More details on authentication with Kiota can be found in the [Kiota authentication documentation](/openapi/kiota/a ...
+
+What to remember is that the `GitHubClientFactory` is used through dependency injection, which means you can have other types injected by the constructor that are needed for the authentication provider.
```
```suggestion
The `GitHubClientFactory` is used through dependency injection, which means you can have other types injected by the constructor that are needed for the authentication provider.
```

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependencyinjection.md:
```diff
@@ -0,0 +1,288 @@
+
+What to remember is that the `GitHubClientFactory` is used through dependency injection, which means you can have other types injected by the constructor that are needed for the authentication provider.
+
+In [this example](https://github.com/svrooij/KiotaWithDependencyInjection/blob/d1ee972e8ba06c4f9707b78dfa981f5f57fbf7e6/KiotaWithDependencyInjection/Spotify/SpotifyFactory.cs) there is an `IAccessTokenProvider` injected that is then [used](https://github.com/svrooij/KiotaWithDependencyInjection/blob/d1ee972e8ba06c4f9707b78dfa981f5f57fbf7e6/KiotaWithDependencyInjection/Spotify/SpotifyFactory.cs#L3 ...
```
links to "3rd party repositories" will lead to scrutiny. I suggest removing this section

svrooij replied:
No sample better than me creating a pr for the samples repo?

baywet replied:
we could alternatively do that if you feel like adding more samples to the kiota-samples repo

svrooij replied:
Might create the sample in the sample repo and then create a new pr to update this one.

baywet reviewed (changes requested):
Thanks for this great contribution!
A first round of initial feedback. That plus suggesting moving the sample to kiota samples so it can be referenced here instead of copied.

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependency-injection.md:
```diff
@@ -0,0 +1,301 @@
+
+## Generate the API client
+
+Kiota generates API clients from OpenAPI documents. Create a file named _github-releases.yml_ and add the following.
```
have we considered using the real description from the search results instead and using filters?

svrooij replied:
What do you propose?

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependency-injection.md:
```diff
@@ -0,0 +1,301 @@
+            services.AddTransient(handler);
+        }
+
+        // Or register them manually
+        // services.AddTransient<UriReplacementHandler<UriReplacementHandlerOption>>();
+        // services.AddTransient<RetryHandler>();
+        // services.AddTransient<RedirectHandler>();
+        // services.AddTransient<ParametersNameDecodingHandler>();
+        // services.AddTransient<UserAgentHandler>();
+        // services.AddTransient<HeadersInspectionHandler>();
```
```suggestion
```

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependency-injection.md:
```diff
@@ -0,0 +1,301 @@
+    {
+        // Dynamically load the Kiota handlers from the Client Factory
+        var kiotaHandlers = KiotaClientFactory.GetDefaultHandlerTypes();
+        // And register them in the DI container
```
```suggestion
        // And register them in the dependency injection container
```

baywet commented on OpenAPI/kiota/tutorials/dotnet-dependency-injection.md:
```diff
@@ -0,0 +1,301 @@
+        {
+            builder.AddHttpMessageHandler((sp) => (DelegatingHandler)sp.GetRequiredService(handler));
+        }    
+        
+        // Or attach them manually
+        // builder.AddHttpMessageHandler<UriReplacementHandler<UriReplacementHandlerOption>>();
+        // builder.AddHttpMessageHandler<RetryHandler>();
+        // builder.AddHttpMessageHandler<RedirectHandler>();
+        // builder.AddHttpMessageHandler<ParametersNameDecodingHandler>();
+        // builder.AddHttpMessageHandler<UserAgentHandler>();
+        // builder.AddHttpMessageHandler<HeadersInspectionHandler>();
```
```suggestion

```

baywet reviewed (changes requested):
Thanks for making the updates.
I think we're getting really close!
Are you planning to add the sample in the kiota samples repo so we can use references here instead?

svrooij commented:
> Are you planning to add the sample in the kiota samples repo so we can use references here instead?

Yes, I'll be doing that today. Does it already have a generated client I can use? So I can skip that part?

svrooij commented:
> > Are you planning to add the sample in the kiota samples repo so we can use references here instead?
> 
> Yes, I'll be doing that today. Does it already have a generated client I can use? So I can skip that part?

First hurdle below, solution: `git config --system core.longpaths true`

```plain
Cloning into 'ms-kiota-samples'...
remote: Enumerating objects: 96280, done.
remote: Counting objects: 100% (4770/4770), done.
remote: Compressing objects: 100% (1740/1740), done.
remote: Total 96280 (delta 3228), reused 4443 (delta 2981), pack-reused 91510
Receiving objects: 100% (96280/96280), 23.33 MiB | 20.95 MiB/s, done.
Resolving deltas: 100% (69230/69230), done.
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/Messages/Item/MultiValueExtendedProperties/Item/MultiValueLegacyExtendedPropertyItemRequestBuilderfbdb00d62c7a8f0e2c85cf8cf3ab976d75b4161f94af95af394cded9032b9711.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/Messages/Item/MultiValueExtendedProperties/MultiValueExtendedPropertiesRequestBuilder73077495a2575b048d523418fec9ad4c654063e879cbdce3ddec0500de0bc5a1.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/Messages/Item/SingleValueExtendedProperties/Item/SingleValueLegacyExtendedPropertyItemRequestBuilder25ba096782140d1312334fc503c5a36018b4a791080aa2e0fc674bf1db569203.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/Messages/Item/SingleValueExtendedProperties/SingleValueExtendedPropertiesRequestBuilder40733ae55d40d3e6ddcf10856d0d3f5240a9e3f2d2eb06b4eefa28b0fe72438c.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/MultiValueExtendedProperties/Item/MultiValueLegacyExtendedPropertyItemRequestBuilder5b55a97c5b0d8e76b5af72625258811ae7902b3792ca4f9ebaccab25e7570c93.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/MultiValueExtendedProperties/MultiValueExtendedPropertiesRequestBuilder13e11e0482d79ca042f52ca4b3f44d026c03616a2e5f5096bb9fb6862187e014.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/SingleValueExtendedProperties/Item/SingleValueLegacyExtendedPropertyItemRequestBuilder2686d0ca41ecb14e743a2ff4fa5506bccbd6e801c007ed052916719e94058282.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/ChildFolders/Item/SingleValueExtendedProperties/SingleValueExtendedPropertiesRequestBuilder2395afc9cbd74fb96b1da02a06ca5cd6e272b0ce008816388764df1c74bdf7d1.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/Messages/Item/MultiValueExtendedProperties/Item/MultiValueLegacyExtendedPropertyItemRequestBuilderfa30ff984bb890f890a81a3bc7bd08d649bdb6073054a279850aba7936dd05b3.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/Messages/Item/MultiValueExtendedProperties/MultiValueExtendedPropertiesRequestBuilder154c3234197de6208140c367f7b8a9d1c739ac30c517b3d24a6cb21e0eb4b7be.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/Messages/Item/SingleValueExtendedProperties/Item/SingleValueLegacyExtendedPropertyItemRequestBuilder75d57878e729fc875dc49115cf22144dc3e782f633c1ccc6c65cbd3a0bebb1c6.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/Messages/Item/SingleValueExtendedProperties/SingleValueExtendedPropertiesRequestBuilder2de3f5c488403b53e53405ebf1ebb0e3618f3cc4ccba117f06a613bfbd9775f7.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/MultiValueExtendedProperties/Item/MultiValueLegacyExtendedPropertyItemRequestBuilder6eb08fd81a0441d1c7cef789191817cf5800368a25e7eaec6fa4b4b839b3515f.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/MailFolders/Item/SingleValueExtendedProperties/Item/SingleValueLegacyExtendedPropertyItemRequestBuilderb8a911797ce2a5d481a7353354a030cbe3b5501c5e56ccf3d70654cde0b6b8a6.swift: Filename too long
error: unable to create file msgraph-mail/swiftconsoleapp/Sources/swiftconsoleapp/Users/Item/Messages/Item/SingleValueExtendedProperties/Item/SingleValueLegacyExtendedPropertyItemRequestBuilder077fc46278ec970290bb1f54c2402167a1e4e858a97f0a924c1fc4f1a0ba5cae.swift: Filename too long
fatal: unable to checkout working tree
warning: Clone succeeded, but checkout failed.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'
```

svrooij commented on OpenAPI/kiota/tutorials/dotnet-dependency-injection.md:
````diff
@@ -0,0 +1,136 @@
+
+Open the _Program.cs_ file and add the following code above the `var app = builder.Build();` line. Sample [Program.cs](https://github.com/microsoft/kiota-samples/blob/main/get-started/dotnet-dependency-injection/Program.cs).
+
+```csharp
````
@baywet for clarity this code is still added to the docs, can I highlight specific lines in the included code block?

baywet replied:
I don't know about that, but maybe @jasonjoh can help. The question here is: when referencing a code file with the `:::code` syntax, is there a way to only reference specific lines? or highlight them?

jasonjoh replied:
I will caution against using this heavily though - changes to the code between your snippet markers can cause the line numbers to change, which can throw off the highlighting if you don't remember to update that.

svrooij commented:
@baywet I've changed it to download the github specs with the kiota client and to generated based upon that. Also changed the highlighting part.

Because of all the commits it took, I suggest to Squash this PR

baywet reviewed (approved):
Thank you for the continuous effort on this one it will be a great addition to current documentation.

baywet commented:
Now that #103 has been merged, it should be within the day.
